velox/functions/sparksql/Split.cpp

```diff
@@ -0,0 +1,259 @@
+      arrayWriter.add_item().setNoCopy(input);
+      resultWriter.commit();
+      return;
+    }
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // add them to the elements vector, until we reach the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    auto* re = cache_.findOrCompile(delim);
+    const size_t end = input.size();
+    const char* start = input.data();
+    const auto re2String = re2::StringPiece(start, end);
+    size_t pos = 0;
+
+    re2::StringPiece subMatches[1];
+    // Matches a regular expression against a portion of the input string,
+    // starting from 'pos' to the end of the input string. The match is not
+    // anchored, which means it can start at any position in the string. If a
+    // match is found, the matched portion of the string is stored in
+    // 'subMatches'. The '1' indicates that we are only interested in the first
+    // match found from the current position 'pos' in each iteration of the
+    // loop.
+    while (re->Match(
+        re2String, pos, end, RE2::Anchor::UNANCHORED, subMatches, 1)) {
+      const auto fullMatch = subMatches[0];
+      auto offset = fullMatch.data() - start;
+      const auto size = fullMatch.size();
+      if (offset >= end) {
+        break;
+      }
+
+      // When hitting an empty match, split the character at the current 'pos'
+      // of the input string and put it into the result array, followed by an
+      // empty tail string at last, e.g., the result array for split('abc','d|')
+      // is ["a", "b", "c",""].
+      if (size == 0) {
+        offset += tryGetCharLength(start + pos, end - pos);
+      }
+      arrayWriter.add_item().setNoCopy(StringView(start + pos, offset - pos));
+      pos = offset + size;
+
+      ++addedElements;
+      // If the next element should be the last, leave the loop.
+      if (addedElements + 1 == limit) {
+        break;
+      }
+    }
+
+    // Add the rest of the string and we are done.
+    // Note that the rest of the string can be empty - we still add it.
+    arrayWriter.add_item().setNoCopy(StringView(start + pos, end - pos));
+    resultWriter.commit();
+  }
+
+  mutable functions::detail::ReCache cache_;
+};
+
+std::shared_ptr<exec::VectorFunction> createSplit(
+    const std::string& /*name*/,
+    const std::vector<exec::VectorFunctionArg>& inputArgs,
+    const core::QueryConfig& /*config*/) {
+  VELOX_USER_CHECK(
+      inputArgs.size() == 2 || inputArgs.size() == 3,
+      "Two or three arguments are required for split function.");
+  VELOX_USER_CHECK(
+      inputArgs[0].type->isVarchar(),
+      "The first argument should be of varchar type.");
+  VELOX_USER_CHECK(
+      inputArgs[1].type->isVarchar(),
+      "The second argument should be of varchar type.");
+  if (inputArgs.size() > 2) {
+    VELOX_USER_CHECK(
+        inputArgs[2].type->kind() == TypeKind::INTEGER,
+        "The third argument should be of integer type.");
```

Comment:
I actually have same thought, but i remember per rui's ask. let wait @rui-mo 's input.

Reply:
@PHILO-HE @gaoyangxiaozhu Thanks. I agree. This is not needed for this function. Perhaps remove this method and declare it as vector function instead of **stateful** vector function as below.

```
VELOX_DECLARE_VECTOR_FUNCTION(
    udf_regexp_split,
    signatures(),
    std::make_unique<Split>());
```